### PR TITLE
Speedup calculate_tx_index_from_block

### PR DESCRIPTION
**Pull request by pavel-kokolemin**
Calculate each transaction offset just once.

### CODE REVIEW

File: common/src/chain/transaction/transaction_index/mod.rs [hunk 1]
```diff
@@ -144,40 +144,23 @@ pub struct TxMainChainIndex {
...
-
-    TxMainChainIndex::new(
-        SpendablePosition::from(tx_position),
-        tx.outputs()
-            .len()
-            .try_into()
-            .expect("Number conversion from usize to u32 should not fail here (3)"),
-    )
+pub fn calculate_tx_offsets_in_block(block: &Block) -> Vec<u32> {
+    // Please add new test if new block format version is added
```
Comment (TheQuantumPhysicist) [inline common/src/chain/transaction/transaction_index/mod.rs:148]: You don't need to add this comment. The match below will yell at whoever tries to ignore this ;-)

File: chainstate/tx-verifier/src/transaction_verifier/tx_index_cache.rs [hunk 2]
```diff
@@ -48,10 +48,24 @@ impl TxIndexCache {
         self.data
     }
 
-    pub fn add_tx_index(&mut self, spend_ref: BlockTransactableRef) -> Result<(), TxIndexError> {
+    pub fn add_tx_index(
+        &mut self,
+        spend_ref: BlockTransactableRef,
+        tx_offsets: &[u32],
+    ) -> Result<(), TxIndexError> {
         let tx_index = match spend_ref {
```
Comment (TheQuantumPhysicist) [inline chainstate/tx-verifier/src/transaction_verifier/tx_index_cache.rs:56]: Please debug_assert here that the lengths are equal.

**Review by TheQuantumPhysicist — APPROVED** [review]
Besides the minor things I mentioned, all good :-)

File: chainstate/tx-verifier/src/transaction_verifier/tx_index_cache.rs (hunk 2 above)
Comment (pavel-kokolemin) [reply chainstate/tx-verifier/src/transaction_verifier/tx_index_cache.rs:56]: Added, thanks.

File: common/src/chain/transaction/transaction_index/mod.rs (hunk 1 above)
Comment (TheQuantumPhysicist) [reply common/src/chain/transaction/transaction_index/mod.rs]: The better way is to create a superficial match there that triggers an error :wink:

**Review by stanislav-tkach — APPROVED** [review]

File: common/src/chain/transaction/transaction_index/mod.rs [hunk 3]
```diff
@@ -144,40 +144,21 @@ pub struct TxMainChainIndex {
...
-    let tx_position = TxMainChainPosition::new(block.get_id(), offset_tx);
-
-    TxMainChainIndex::new(
-        SpendablePosition::from(tx_position),
-        tx.outputs()
-            .len()
-            .try_into()
-            .expect("Number conversion from usize to u32 should not fail here (3)"),
-    )
+pub fn calculate_tx_offsets_in_block(block: &Block) -> Vec<u32> {
```
Comment (azarovh) [inline common/src/chain/transaction/transaction_index/mod.rs:147]: why not returning `Vec<TxMainChainIndex>`? Having raw offsets in u32 form all over the code is not the best imo

Comment (TheQuantumPhysicist) [reply common/src/chain/transaction/transaction_index/mod.rs:147]: This sounds like a better solution. Pavel, do you mind trying it out?

**Comment by TheQuantumPhysicist** [comment]
There's actually an issue I noticed now. It's that we have to calculate and pass the offset even if we don't need it because the transaction index is disabled... if we want to calculate it only if we need it, then we have to uphold that invariant, which is ugly...

I don't know what the best solution could be.

File: common/src/chain/transaction/transaction_index/mod.rs (hunk 3 above)
Comment (TheQuantumPhysicist) [reply common/src/chain/transaction/transaction_index/mod.rs:147]: Now I also noticed that if we pass the index by reference, we'll have to copy it. It seems that the reasonable solution is to pass the tx index of a transaction while connecting it, not the whole vector as a slice.

Comment (pavel-kokolemin) [reply common/src/chain/transaction/transaction_index/mod.rs]: > This sounds like a better solution. Pavel, do you mind trying it out?

Sure, updated.

> Now I also noticed that if we pass the index by reference, we'll have to copy it. It seems that the reasonable solution is to pass the tx index of a transaction while connecting it, not the whole vector as a slice.

Updated. Looks better now, thanks.

**Comment by pavel-kokolemin** [comment]
> There's actually an issue I noticed now. It's that we have to calculate and pass the offset even if we don't need it because the transaction index is disabled... if we want to calculate it only if we need it, then we have to uphold that invariant, which is ugly...
> I don't know what the best solution could be.

Do I need to make `calculate_tx_offsets_in_block` call optional? It's indeed a bit ugly.

File: chainstate/tx-verifier/src/transaction_verifier/mod.rs [hunk 4]
```diff
@@ -700,11 +720,11 @@ impl<'a, S: TransactionVerifierStorageRef, U: UtxosView> TransactionVerifier<'a,
     pub fn connect_transactable(
         &mut self,
         block_index: &BlockIndex,
-        spend_ref: BlockTransactableRef,
+        spend_ref: BlockTransactableWithIndexRef,
```
[the remark this answers is not in the document]
Comment (azarovh) [reply chainstate/tx-verifier/src/transaction_verifier/mod.rs:723]: sorry, won't work, never mind

**Comment by TheQuantumPhysicist** [comment]
I revamped the whole solution around TransactionVerifierStrategy. Besides the unnecessary cloning of the transaction index, the primary fear is to have inconsistent state of the transaction index taken from the configuration. If transaction index is disabled, there's no need to calculate the transaction indices. Holding this invariant consistently is the goal here.

**Comment by pavel-kokolemin** [comment]
> I revamped the whole solution around TransactionVerifierStrategy. Besides the unnecessary cloning of the transaction index, the primary fear is to have inconsistent state of the transaction index taken from the configuration. If transaction index is disabled, there's no need to calculate the transaction indices. Holding this invariant consistently is the goal here.

Looks good, thanks.

File: chainstate/src/detail/tx_verification_strategy/tx_verification_strategy_utils.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+use common::{
+    chain::{calculate_tx_offsets_in_block, Block, TxMainChainIndex, TxMainChainIndexError},
+    primitives::Idable,
+};
+use tx_verifier::transaction_verifier::{config::TransactionVerifierConfig, error::TxIndexError};
+
+pub fn take_tx_index(
+    tx_indices: &mut Option<VecDeque<TxMainChainIndex>>,
+) -> Option<TxMainChainIndex> {
+    match tx_indices {
```
Comment (azarovh) [inline chainstate/src/detail/tx_verification_strategy/tx_verification_strategy_utils.rs:27]:
```rust
tx_indices.map(|v| v.pop_from())
```

File: chainstate/src/detail/tx_verification_strategy/tx_verification_strategy_utils.rs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+use std::collections::VecDeque;
+
+use common::{
+    chain::{calculate_tx_offsets_in_block, Block, TxMainChainIndex, TxMainChainIndexError},
+    primitives::Idable,
+};
+use tx_verifier::transaction_verifier::{config::TransactionVerifierConfig, error::TxIndexError};
+
+pub fn take_tx_index(
```
Comment (azarovh) [inline chainstate/src/detail/tx_verification_strategy/tx_verification_strategy_utils.rs:24]: `take_front_tx_index`?

**Review by azarovh — APPROVED** [review]

File: chainstate/src/detail/tx_verification_strategy/tx_verification_strategy_utils.rs (hunk 5 above)
Comment (TheQuantumPhysicist) [reply chainstate/src/detail/tx_verification_strategy/tx_verification_strategy_utils.rs:27]: map can't take a mutable reference.

File: chainstate/src/detail/tx_verification_strategy/tx_verification_strategy_utils.rs (hunk 6 above)
Comment (TheQuantumPhysicist) [reply chainstate/src/detail/tx_verification_strategy/tx_verification_strategy_utils.rs:24]: Fair

**Review by stanislav-tkach — APPROVED** [review]

**Pull request merged**